python/demo/mixed-elasticity-sc/static-condensation-elasticity.py

```diff
@@ -0,0 +1,175 @@
+    sigma = E / (1. - nu ** 2) * ((1. - nu) * eps + nu * ufl.Identity(2) * ufl.tr(eps))
+    return sigma
+
+
+a00 = ufl.inner(sigma, sigmat) * ufl.dx
+a10 = - ufl.inner(sigma, ufl.grad(ut)) * ufl.dx
+a01 = - ufl.inner(sigma_u(u), sigmat) * ufl.dx
+
+f = ufl.as_vector([0.0, 1.0 / 16])
+b1 = - ufl.inner(f, ut) * ds(1)
+
+# JIT compile individual blocks tabulation kernels
+ufc_form00 = dolfin.jit.ffc_jit(a00)
+kernel00 = ufc_form00.create_cell_integral(-1).tabulate_tensor
+
+ufc_form01 = dolfin.jit.ffc_jit(a01)
+kernel01 = ufc_form01.create_cell_integral(-1).tabulate_tensor
+
+ufc_form10 = dolfin.jit.ffc_jit(a10)
+kernel10 = ufc_form10.create_cell_integral(-1).tabulate_tensor
+
+ffi = cffi.FFI()
+
+numba.cffi_support.register_type(ffi.typeof('double _Complex'),
+                                 numba.types.complex128)
+
+c_signature = numba.types.void(
+    numba.types.CPointer(numba.typeof(PETSc.ScalarType())),
+    numba.types.CPointer(numba.typeof(PETSc.ScalarType())),
+    numba.types.CPointer(numba.types.double),
+    numba.types.CPointer(numba.types.int32),
+    numba.types.CPointer(numba.types.int32))
+
+
+@numba.cfunc(c_signature, nopython=True)
+def tabulate_condensed_tensor_A(A_, w_, coords_, entity_local_index, cell_orientation):
+    # Prepare target condensed local elem tensor
+    A = numba.carray(A_, (Usize, Usize), dtype=PETSc.ScalarType)
+
+    # Tabulate all sub blocks locally
+    A00 = numpy.zeros((Ssize, Ssize), dtype=PETSc.ScalarType)
+    kernel00(ffi.from_buffer(A00), w_, coords_, entity_local_index, cell_orientation)
+
+    A01 = numpy.zeros((Ssize, Usize), dtype=PETSc.ScalarType)
+    kernel01(ffi.from_buffer(A01), w_, coords_, entity_local_index, cell_orientation)
+
+    A10 = numpy.zeros((Usize, Ssize), dtype=PETSc.ScalarType)
+    kernel10(ffi.from_buffer(A10), w_, coords_, entity_local_index, cell_orientation)
+
+    # A = - A10 * A00^{-1} * A01
+    A00inv = numpy.linalg.inv(A00)
+    A[:, :] = - numpy.dot(A10, numpy.dot(A00inv, A01))
+
+
+# Prepare an empty Form and set the condensed tabulation kernel
+a_cond = dolfin.cpp.fem.Form([U._cpp_object, U._cpp_object])
+a_cond.set_tabulate_cell(-1, tabulate_condensed_tensor_A.address)
+
+A_cond = dolfin.fem.assemble_matrix(a_cond, [bc])
+A_cond.assemble()
+
+b = dolfin.fem.assemble_vector(b1)
+dolfin.fem.apply_lifting(b, [a_cond], [[bc]])
+b.ghostUpdate(addv=PETSc.InsertMode.ADD, mode=PETSc.ScatterMode.REVERSE)
+dolfin.fem.set_bc(b, [bc])
+
+uc = dolfin.Function(U)
+dolfin.la.solve(A_cond, uc.vector(), b)
+
+with dolfin.io.XDMFFile(dolfin.MPI.comm_world, "uc.xdmf") as outfile:
+    outfile.write_checkpoint(uc, "uc")
+
+# Pure displacement based formulation
+a = - ufl.inner(sigma_u(u), ufl.grad(ut)) * ufl.dx
+A = dolfin.fem.assemble_matrix(a, [bc])
+A.assemble()
+
+# Extract bounding box for function evaluation
+bb_tree = dolfin.cpp.geometry.BoundingBoxTree(mesh, 2)
+
+# Evaluate at free end midpoint
+# Keep False value in non-owning processes
```

Comment:
Not sure what this comment means.

Reply:
I changed the code for parallel evaluation.